File: src/1135-Connecting-Cities-With-Minimum-Cost/1135.js
var minimumCost = function(N, connections) {
    if (connections.length < N - 1) return -1;
    connections.sort(function(a, b){
        return a[2] - b[2];
    });
    
    var parent = Array(N);
    for (var i = 0; i < N; ++i) parent[i] = i;
    
    function find(x) {
        if (x != parent[x]) parent[x] = find(parent[x]);
        return parent[x];
    }
    
    var e = 0, res = 0, k = 0;
    while (e < N - 1) {
        var pre = connections[k++];
        var x = find(pre[0] - 1), y = find(pre[1] - 1);
        if (x != y) {
            parent[x] = y;
            e++;
            res += pre[2];
        }
    }
    return res;
};

File: src/1135-Connecting-Cities-With-Minimum-Cost/1135.cpp
class Solution 
{
public:
    int minimumCost(int N, vector<vector<int>>& connections) 
    {
        if (connections.size() < N - 1) return -1;
        for (int i = 0; i < N; ++i) parent.push_back(i);
        sort(connections.begin(), connections.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[2] < b[2];
        });
        int e = 0, res = 0, k = 0;
        while (e < N - 1)
        {
            auto pre = connections[k++];
            int x = find(pre[0] - 1), y = find(pre[1] - 1);
            if (x != y)
            {
                parent[x] = y;
                e++;
                res += pre[2];
            }
        }
        return res;
    }
    
    vector<int> parent;
    int find(int x)
    {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
};